src/main/java/net/neoforged/neoforge/server/command/generation/GenerationTask.java

```diff
@@ -0,0 +1,248 @@
+ */
+public class GenerationTask {
+    private static final Logger LOGGER = LogManager.getLogger();
+    private static final int BATCH_SIZE = 32;
+    private static final int QUEUE_THRESHOLD = 8;
+    private static final int COARSE_CELL_SIZE = 4;
+
+    private final MinecraftServer server;
+    private final ServerChunkCache chunkSource;
+
+    private final Iterator<ChunkPos> iterator;
+    private final int x;
+    private final int z;
+    private final int radius;
+
+    private final int totalCount;
+
+    private final Object queueLock = new Object();
+
+    private final AtomicInteger queuedCount = new AtomicInteger();
+    private final AtomicInteger okCount = new AtomicInteger();
+    private final AtomicInteger errorCount = new AtomicInteger();
+    private final AtomicInteger skippedCount = new AtomicInteger();
+
+    private volatile Listener listener;
+    private volatile boolean stopped;
+
+    public static final TicketType<ChunkPos> NEOFORGE_GENERATE_FORCED = TicketType.create("neoforge_generate_forced", Comparator.comparingLong(ChunkPos::toLong));
+
+    public GenerationTask(ServerLevel serverLevel, int x, int z, int radius) {
+        this.server = serverLevel.getServer();
+        this.chunkSource = serverLevel.getChunkSource();
+
+        this.iterator = new CoarseOnionIterator(radius, COARSE_CELL_SIZE);
+        this.x = x;
+        this.z = z;
+        this.radius = radius;
+
+        int diameter = radius * 2 + 1;
+        this.totalCount = diameter * diameter;
+    }
+
+    public int getOkCount() {
+        return this.okCount.get();
+    }
+
+    public int getErrorCount() {
+        return this.errorCount.get();
+    }
+
+    public int getSkippedCount() {
+        return this.skippedCount.get();
+    }
+
+    public int getTotalCount() {
+        return this.totalCount;
+    }
+
+    public void run(Listener listener) {
+        if (this.listener != null) {
+            throw new IllegalStateException("already running!");
+        }
+
+        this.listener = listener;
+
+        this.server.submit(this::skipInitialGeneratedChunksAndStartGenAfterwards);
+    }
+
+    private void skipInitialGeneratedChunksAndStartGenAfterwards() {
+        CompletableFuture.runAsync(() -> {
+            Iterator<ChunkPos> iterator = this.iterator;
+            while (iterator.hasNext()) {
+                ChunkPos chunkPosInLocalSpace = iterator.next();
+                if (Math.abs(chunkPosInLocalSpace.x) <= this.radius && Math.abs(chunkPosInLocalSpace.z) <= this.radius) {
+                    if (isChunkFullyGenerated(chunkPosInLocalSpace)) {
+                        this.skippedCount.incrementAndGet();
+                        this.listener.update(this.okCount.get(), this.errorCount.get(), this.skippedCount.get(), this.totalCount);
+                        continue;
+                    }
+
+                    this.tryEnqueueTasks();
+                    break;
+                }
+            }
+        });
```

Comment:
You should maybe put a different threadpool into this task because otherwise this is running on the Server thread.
And if there is enough chunks to skip this could lead to a server kill.